File: patterns/04-concurrency-patterns/05-future-promise/python/future_promise.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
def evaluate(input_data):
    state="pending";value="none";error="none";observations=[];terminal_count=0
    for action in input_data.get("actions",[]):
        op=action["op"]
        if op in ("complete","fail","cancel"):
            if state!="pending":raise PatternError("already_completed")
            terminal_count+=1
            if op=="complete":state="fulfilled";value=action["value"]
            elif op=="fail":state="rejected";error=action["code"]
            else:state="cancelled"
        elif op=="observe": observations.append("pending" if state=="pending" else f"value:{value}" if state=="fulfilled" else f"error:{error}" if state=="rejected" else "cancelled")
        else:raise PatternError("unknown_action")
    return {"state":state,"observations":observations,"value":value,"error":error,"terminal_count":terminal_count}

File: patterns/04-concurrency-patterns/05-future-promise/go/future_promise.go
package futurepromise

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }
func Evaluate(input map[string]any) (any, error) {
	state, value, errorValue := "pending", "none", "none"
	observations := []string{}
	terminal := 0
	actions, _ := input["actions"].([]any)
	for _, raw := range actions {
		action := raw.(map[string]any)
		op := action["op"].(string)
		switch op {
		case "complete", "fail", "cancel":
			if state != "pending" {
				return nil, &PatternError{code: "already_completed"}
			}
			terminal++
			if op == "complete" {
				state = "fulfilled"
				value = action["value"].(string)
			} else if op == "fail" {
				state = "rejected"
				errorValue = action["code"].(string)
			} else {
				state = "cancelled"
			}
		case "observe":
			if state == "pending" {
				observations = append(observations, "pending")
			} else if state == "fulfilled" {
				observations = append(observations, "value:"+value)
			} else if state == "rejected" {
				observations = append(observations, "error:"+errorValue)
			} else {
				observations = append(observations, "cancelled")
			}
		default:
			return nil, &PatternError{code: "unknown_action"}
		}
	}
	return map[string]any{"state": state, "observations": observations, "value": value, "error": errorValue, "terminal_count": terminal}, nil
}
